### PR TITLE
 [wip]Fix low accuracy of Qwen3.5-27B

### PR DESCRIPTION
**Pull request by ccrhx4**
This PR is to fix low accuracy of Qwen3.5-27B on B60. [VLLMZ-1515](https://jira.devtools.intel.com/browse/VLLMZ-1515)

Before:
vllm ({'pretrained': '/hf_models/Qwen3.5-27B/', 'enforce_eager': True, 'tensor_parallel_size': 4, 'add_bos_token': True, 'max_model_len': 32768}), gen_kwargs: ({'max_gen_toks': 4096}), limit: 200.0, num_fewshot: 5, batch_size: auto
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.085|±  |0.0198|
|     |       |strict-match    |     5|exact_match|↑  |0.080|±  |0.0192|

vllm ({'pretrained': '/hf_models/Qwen3.6-27B/', 'enforce_eager': True, 'tensor_parallel_size': 4, 'add_bos_token': True}), gen_kwargs: ({'max_gen_toks': 4096}), limit: 200.0, num_fewshot: 5, batch_size: auto
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |    0|±  |     0|
|     |       |strict-match    |     5|exact_match|↑  |    0|±  |     0|

After:
vllm ({'pretrained': '/hf_models/Qwen3.5-27B/', 'enforce_eager': True, 'tensor_parallel_size': 4, 'add_bos_token': True}), gen_kwargs: ({'max_gen_toks': 4096}), limit: 200.0, num_fewshot: 5, batch_size: auto
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.865|±  |0.0242|
|     |       |strict-match    |     5|exact_match|↑  |0.880|±  |0.0230|

vllm ({'pretrained': '/hf_models/Qwen3.6-27B/', 'enforce_eager': True, 'tensor_parallel_size': 4, 'add_bos_token': True}), gen_kwargs: ({'max_gen_toks': 4096}), limit: 200.0, num_fewshot: 5, batch_size: auto
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.700|±  |0.0325|
|     |       |strict-match    |     5|exact_match|↑  |0.705|±  |0.0323|

FULL GSM8K Test Result:
vllm ({'pretrained': '/hf_models/Qwen3.5-27B/', 'enforce_eager': True, 'tensor_parallel_size': 4, 'add_bos_token': True}), gen_kwargs: ({'max_gen_toks': 4096}), limit: None, num_fewshot: 5, batch_size: auto
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.8696|±  |0.0093|
|     |       |strict-match    |     5|exact_match|↑  |0.8878|±  |0.0087|

### CODE REVIEW

**Comment by ccrhx4** [comment]
@YangQun1 @pengzhao-intel @mayuyuace Please help to review.

**Comment by YangQun1** [comment]
Thanks for the fixes. There are fixes related to several files/kernels. Could you explain the rationale behind each fix, and if all of them are necessary for accuracy?

File: csrc/xpu/gdn_attn/xe_2/chunk_gated_delta_rule_kernels_xe2.hpp [hunk 1]
```diff
@@ -378,6 +378,7 @@ CUTE_DEVICE void chunk_inverse_kernel(
                 static_cast<T>(A_ptr_save[m_idx * chunk_size + n_idx]);
           }
         }
+        item.barrier(sycl::access::fence_space::local_space);
```
Comment (YangQun1) [inline csrc/xpu/gdn_attn/xe_2/chunk_gated_delta_rule_kernels_xe2.hpp:381]: the chunk_inverse_kernel should only be used on PVC

Comment (ccrhx4) [reply csrc/xpu/gdn_attn/xe_2/chunk_gated_delta_rule_kernels_xe2.hpp:381]: Yes, the code is patched here for PVC.

Comment (YangQun1) [reply csrc/xpu/gdn_attn/xe_2/chunk_gated_delta_rule_kernels_xe2.hpp:381]: why it impact the B60 acc?

Comment (ccrhx4) [reply csrc/xpu/gdn_attn/xe_2/chunk_gated_delta_rule_kernels_xe2.hpp:381]: I patched for PVC code as the code is there for PVC, although I did not test on PVC.

**Comment by mayuyuace** [comment]
LGTM.
@pengzhao-intel
Please help review code changes of chunk_causal_conv1d_tiled_xe2.hpp

File: csrc/xpu/gdn_attn/xe_2/chunk_gated_delta_rule_kernels_xe2.hpp (hunk 1 above)
Comment (mayuyuace) [reply csrc/xpu/gdn_attn/xe_2/chunk_gated_delta_rule_kernels_xe2.hpp:381]: Barrier in for-loop may affect kernel performance.

Comment (YangQun1) [reply csrc/xpu/gdn_attn/xe_2/chunk_gated_delta_rule_kernels_xe2.hpp:381]: Ok, just a little bit confused. I saw the pr target is to fix B60 acc. if this change is not for B60, suggest removing it for now. if it's to fix acc for PVC and it's verified, we can keep it and minor update the PR description.

**Comment by ccrhx4** [comment]
The fix follows commit 22c60d69a083ab8d989e5061b93944e095f32bae
As the Qwen3.5/3.6-27B models have a different kv head number than Qwen3.5-9B and 35B-A3B. It triggers the bug.

 For Qwen3.5-27B, the GDN layers have the following dimensions:
   - num_k_heads = 16, num_v_heads = 48
   - head_k_dim = 128, head_v_dim = 128
   - Total convolution elements (conv_elems): q_dim + k_dim + v_dim = $(16 \times 128) + (16 \times 128) + (48 \times 128) = 10,240$ elements.

 When running with Tensor Parallelism (tp_size = 4), the heads are split across ranks:
   - conv_elems per TP rank: $10,240 \div 4 =$ 2,560 elements.
  The state-update kernel is configured with group_size = 256 and elems_per_item = 4, meaning each work-group processes 1,024 elements (elems_per_group).

   - 2,560 is not a multiple of 1,024 ($2,560 \div 1,024 = 2.5$). 
   - Three workgroups are launched. The third work-group processes elements up to index 3,072, leaving it over-provisioned by 512 elements.

  Impact: Without the explicit if (i < conv_elems) check, the third work-group writes 512 out-of-bounds elements into adjacent PyTorch tensor allocations on every layer, leading to catastrophic state corruption, random generation behavior, and garbage outputs. The bounds check fully prevents this memory overflow.

File: csrc/xpu/gdn_attn/xe_2/chunk_gated_delta_rule_kernels_xe2.hpp (hunk 1 above)
Comment (ccrhx4) [reply csrc/xpu/gdn_attn/xe_2/chunk_gated_delta_rule_kernels_xe2.hpp:381]: Hi Qun, I take your side, I will close this PR for now and have it tested on PVC first.

**Comment by ccrhx4** [comment]
found some regression in the test. close it for now.
